### DISCUSSION TITLE
pnpmfile should not be ignored

### DISCUSSION THREAD

**Discussion by Djaler** [renovatebot/renovate#16063] [Suggest an Idea]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I have a pnpmfile in my project. It is required to fix some dependencies issues within monorepo.
You can see it here https://github.com/Djaler/vue-use-route-query/blob/master/.pnpmfile.cjs

But renovate ignores the pnpmfile, which causes incorrect dependencies resolving for me (you can see here that vue-template-compiler version was changed https://github.com/Djaler/vue-use-route-query/pull/17/files)

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Generating pnpm-lock.yaml for .(branch="renovate/simple-git-hooks-2.x")
DEBUG: Spawning pnpm install to create pnpm-lock.yaml(branch="renovate/simple-git-hooks-2.x")
DEBUG: No node constraint found - using latest(branch="renovate/simple-git-hooks-2.x")
DEBUG: pnpm command(branch="renovate/simple-git-hooks-2.x")
{
  "cmd": "pnpm",
  "args": "install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile"
}
DEBUG: Using docker to execute(branch="renovate/simple-git-hooks-2.x")
{
  "image": "node"
}
DEBUG: No tag or tagConstraint specified(branch="renovate/simple-git-hooks-2.x")
{
  "image": "docker.io/renovate/node"
}
DEBUG: Docker image is already prefetched: docker.io/renovate/node@sha256:4a5086462974dc38d4ff70871ac59429d0acde33eb5b9fb33aa3a9d892b47292(branch="renovate/simple-git-hooks-2.x")
DEBUG: Executing command(branch="renovate/simple-git-hooks-2.x")
{
  "command": "docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/Djaler/vue-use-route-query\":\"/mnt/renovate/gh/Djaler/vue-use-route-query\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/mnt/renovate/gh/Djaler/vue-use-route-query\" docker.io/renovate/node bash -l -c \"install-tool pnpm 7.1.6 && pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\""
}
DEBUG: exec completed(branch="renovate/simple-git-hooks-2.x")
{
  "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/Djaler/vue-use-route-query\":\"/mnt/renovate/gh/Djaler/vue-use-route-query\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/mnt/renovate/gh/Djaler/vue-use-route-query\" docker.io/renovate/node bash -l -c \"install-tool pnpm 7.1.6 && pnpm install --recursive --lockfile-only --ignore-scripts --ignore-pnpmfile\"",
  "durationMs": 47013,
  "stdout": "Installing legacy tool pnpm v7.1.6\n\nadded 1 package in 3s\n7.1.6\nScope: all 6 workspace projects\nProgress: resolved 1, reused 0, downloaded 0, added 0\nProgress: resolved 13, reused 0, downloaded 7, added 0\nProgress: resolved 18, reused 0, downloaded 11, added 0\nProgress: resolved 21, reused 0, downloaded 16, added 0\nProgress: resolved 23, reused 0, downloaded 20, added 0\nProgress: resolved 25, reused 0, downloaded 22, added 0\nProgress: resolved 40, reused 0, downloaded 36, added 0\nProgress: resolved 66, reused 0, downloaded 55, added 0\nProgress: resolved 85, reused 0, downloaded 77, added 0\npackages/core                            |  WARN  deprecated rollup-plugin-virtual@1.0.1\nProgress: resolved 100, reused 0, downloaded 89, added 0\nProgress: resolved 119, reused 0, downloaded 110, added 0\nProgress: resolved 136, reused 0, downloaded 125, added 0\nProgress: resolved 150, reused 0, downloaded 140, added 0\nProgress: resolved 163, reused 0, downloaded 159, added 0\nProgress: resolved 182, reused 0, downloaded 176, added 0\nProgress: resolved 217, reused 0, downloaded 206, added 0\nProgress: resolved 240, reused 0, downloaded 230, added 0\nProgress: resolved 265, reused 0, downloaded 236, added 0\nProgress: resolved 286, reused 0, downloaded 251, added 0\nProgress: resolved 296, reused 0, downloaded 273, added 0\nProgress: resolved 311, reused 0, downloaded 288, added 0\nProgress: resolved 311, reused 0, downloaded 289, added 0\nProgress: resolved 311, reused 0, downloaded 290, added 0\nProgress: resolved 311, reused 0, downloaded 291, added 0\nProgress: resolved 312, reused 0, downloaded 291, added 0\nProgress: resolved 341, reused 0, downloaded 316, added 0\nProgress: resolved 378, reused 0, downloaded 345, added 0\nProgress: resolved 432, reused 0, downloaded 392, added 0\nProgress: resolved 477, reused 0, downloaded 441, added 0\nProgress: resolved 524, reused 0, downloaded 490, added 0\nProgress: resolved 572, reused 0, downloaded 534, added 0\nProgress: resolved 614, reused 0, downloaded 585, added 0\nProgress: resolved 626, reused 0, downloaded 599, added 0\nProgress: resolved 634, reused 0, downloaded 608, added 0\nProgress: resolved 646, reused 0, downloaded 618, added 0\nProgress: resolved 660, reused 0, downloaded 634, added 0\nProgress: resolved 675, reused 0, downloaded 649, added 0\n WARN  Issues with peer dependencies found\n.\n├─┬ @djaler/eslint-config-typescript\n│ ├─┬ @djaler/eslint-config-base\n│ │ └─┬ eslint-plugin-import\n│ │   ├── ✕ missing peer @typescript-eslint/parser@\"*\"\n│ │   └─┬ eslint-module-utils\n│ │     └── ✕ missing peer @typescript-eslint/parser@\"*\"\n│ ├─┬ @typescript-eslint/eslint-plugin\n│ │ └── ✕ missing peer @typescript-eslint/parser@^4.0.0\n│ └─┬ eslint-config-airbnb-typescript\n│   └─┬ eslint-config-airbnb\n│     ├── ✕ missing peer eslint-plugin-import@^2.22.1\n│     ├── ✕ missing peer eslint-plugin-jsx-a11y@^6.4.1\n│     ├── ✕ missing peer eslint-plugin-react@^7.21.5\n│     ├── ✕ missing peer eslint-plugin-react-hooks@\"^4 || ^3 || ^2.3.0 || ^1.7.0\"\n│     └─┬ eslint-config-airbnb-base\n│       └── ✕ missing peer eslint-plugin-import@^2.22.1\n└─┬ eslint-import-resolver-typescript\n  └── ✕ missing peer eslint-plugin-import@\"*\"\nPeer dependencies that should be installed:\n  @typescript-eslint/parser@\">=4.0.0 <5.0.0\"\n  eslint-plugin-import@\">=2.22.1 <3.0.0\"\n  eslint-plugin-jsx-a11y@^6.4.1\n  eslint-plugin-react-hooks@\"^4 || ^3 || ^2.3.0 || ^1.7.0\"\n  eslint-plugin-react@^7.21.5\n\nProgress: resolved 675, reused 0, downloaded 655, added 0, done\n",
  "stderr": ""
}
DEBUG: pnpm-lock.yaml needs updating(branch="renovate/simple-git-hooks-2.x")
DEBUG: Updated 1 lock files(branch="renovate/simple-git-hooks-2.x")
{
  "updatedArtifacts": [
    "pnpm-lock.yaml"
  ]
}
```

</details>

### Have you created a minimal reproduction repository?

No reproduction, but I have linked to a public repo where it occurs

**Comment by rarkins** [comment]
As shown in the logs, Renovate installs and calls pnpm to update the lock file. Renovate is therefore not updating the lock file directly and it's therefore not ignoring anything. I don't currently see how this is a bug from Renovate. Does the problem occur only when updating a related dependency?

**Comment by Djaler** [comment]
The problem is pnpm called with `--ignore-pnpmfile` which makes pnpm-lock.yaml different from usual run `pnpm install` on local machine

**Comment by rarkins** [comment]
Thanks for clarifying. We should try to work out why that was added in the first place and if it's safe to remove it completely or if it needs to be a config option

**Comment by PhilipAbed** [comment]
i believe this can be controlled by config PSB : https://github.com/renovatebot/renovate/blob/789cf38d8b416bd3f79fdabacd7aa2334fe23e49/lib/modules/manager/npm/post-update/pnpm.ts#L60-L63

@Djaler  please check if you can change those configs and confirm if it solves your issue

**Comment by PhilipAbed** [comment]
i'm not sure how scripts are connected to the `pnpmFile`, but you can change that with the flags:
the `ignoreScripts` = false by default
the `allowscripts ` = false by default, 

so you need to set `allowscripts` to `true`

https://docs.renovatebot.com/self-hosted-configuration/#allowscripts

**Comment by Djaler** [comment]
I am using Renovate app on github.com, not self-hosted, so I can't configure allowscripts

**Comment by PhilipAbed** [comment]
i think you still can use it in the `config.js` which you point at using the `renovate.json` since its a https://docs.renovatebot.com/getting-started/running/#global-config
@rarkins @viceice  am i right?

**Comment by rarkins** [comment]
OK, so this was added as a security concept, so that it's less easy to run arbitrary code from unknown/untrusted users. But the config in question is clearly not a risk, so that's a challenge.

**Comment by alexeagle** [comment]
I encountered this problem as well, causing a peer dependency to be removed when Renovate updates the pnpm-lock.json file.
https://github.com/aspect-build/rules_js/pull/178/files#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL1888
Indeed it's too bad pnpm doesn't support a structured format to hint extra dependency edges, rather than in-memory arbitrary manipulation of the graph.

**Comment by rarkins** [comment]
I've moved this to an Ideas discussion, as for now it's intentional behavior. A couple of points:

1. Like @alexeagle alluded to, package managers should ideally support all the *dependency* definitions they need in order to better enable automated tooling such as Renovate
2. Once we support better sandboxing support in the hosted app, we should be able to enable this type of unrestricted code anyway

**Comment by Djaler** [comment]
Not a problem solver in a general way, but a workaround for my use case - there is a declarative way to extend package definition with additional information in pnpm: https://pnpm.io/package_json#pnpmpackageextensions
So I have no need in pnpmfile for now https://github.com/Djaler/vue-use-route-query/commit/1afb56f6370d3c6b5413bc0da9d0faa1a6d7aed2

**Reply by viceice** [reply]
we should probably add a note to our doc's for that

**Reply by RIP21** [reply]
If you haven't added it, you probably should :) Also a better error logging would be helpful. In a way, that, if `pnpmfile.js|cjs` is detected, you can emit warning or straight up error. It took me good chunk of time to figure out what's the pitfall of my `pnpm-lock.yaml` being not included in PRs.

**Comment by RIP21** [comment]
Any plans to start supporting/stop ignoring it?
@zkochan maybe you can add more context to this part: 
> OK, so this was added as a security concept, so that it's less easy to run arbitrary code from unknown/untrusted users.

Is it really a security concern? Or `.pnpmfile.(js|cjs)` is somehow protected from some arbitrary code execution and this flag `--ignore-pnpmfile` should just be removed from Renovate default `install` command and everyone will be happy?

This is really a deal breaker for monorepos, frankly speaking, surely pnpm package extensions may help to work around in simple scenarios, but it's too broad sometimes that may cause issues.

**Reply by zkochan** [reply]
I guess renovate doesn't want to run arbitrary code from pnpmfile on its server. You can put anything to a pnpmfile.

In general, it is better to use package extensions or other configurable ways to influence dependency resolution.
